### PR TITLE
Feature: Max length approval programm check for deploy app

### PR DESCRIPTION
**Pull request by rmeena840**
Feature: Max length approval programm check for deploy app.
https://developer.algorand.org/docs/get-details/parameter_tables/?from_query=2048#smart-contract-constraints
The limit is depending on extra pages. The extra page range in [0, 3].

### CODE REVIEW

File: packages/runtime/src/ctx.ts [hunk 1]
```diff
@@ -382,6 +404,10 @@ export class Ctx implements Context {
 			throw new RuntimeError(RUNTIME_ERRORS.GENERAL.INVALID_CLEAR_PROGRAM);
 		}
 
+		const approvalProgramBytes = Buffer.from(approvalProgTEAL, "base64");
+		const clearProgramBytes = Buffer.from(clearProgTEAL, "base64");
```
Comment (thdailong) [inline packages/runtime/src/ctx.ts:408]: the bytes program should be encoded from [algodClient](https://algorand.github.io/js-algorand-sdk/classes/Algodv2.html#compile) 
We also discuss about this one but have not found any idea yet [Discord message](https://discord.com/channels/754286650540032030/754286650540032033/1025398047649308734)

**Review by thdailong — COMMENTED** [review]
I leave some comments
You also need to add changelog and also can you provide some description about this PR please
You could link me to the docs about checking the length of the program too

File: packages/runtime/src/ctx.ts [hunk 2]
```diff
@@ -131,6 +132,27 @@ export class Ctx implements Context {
 		}
 	}
 
+	/**
+	 * Asserts program length
+	 * @param approvalProgramBytes Approval program bytes
+	 * @param clearProgramBytes Clear program bytes
+	 * @param extraPages Extra pages. Default extra page is 0.
+	 */
```
Comment (Megha-Dev-19) [inline packages/runtime/src/ctx.ts:140]: may be we can also add a limit here in description, like we have mentioned errors, so that while adding the method the user can read the docs and realise the limit

File: packages/runtime/test/integration/deploy-app.ts [hunk 3]
```diff
@@ -115,4 +115,148 @@ describe("Algorand Smart Contracts - Stateful Contract Account", function () {
...
+
+		it("Should fail: when program exceeds total allowed program length", function () {
+			const execParams: types.ExecParams = {
+				sign: types.SignType.SecretKey,
+				fromAccount: john.account,
+				type: types.TransactionType.DeployApp,
+				appDefinition: {
+					appName: "Mock",
+					metaType: types.MetaType.FILE,
+					approvalProgramFilename: "very-long-approval.teal", // very large teal program
```
Comment (Megha-Dev-19) [inline packages/runtime/test/integration/deploy-app.ts:152]: the name is quite weird, can't we keep it more sensible, Don't you think.?

File: packages/runtime/test/integration/deploy-app.ts [hunk 4]
```diff
@@ -115,4 +115,148 @@ describe("Algorand Smart Contracts - Stateful Contract Account", function () {
...
+				appDefinition: {
+					appName: "App",
+					metaType: types.MetaType.FILE,
+					approvalProgramFilename: "very-long-approval-2-pages.teal",
+					clearProgramFilename: clearProgramFilename,
+					globalBytes: 1,
+					globalInts: 1,
+					localBytes: 1,
+					localInts: 1,
+					extraPages: 1 // should pass because total 2 pages needed
```
Comment (Megha-Dev-19) [inline packages/runtime/test/integration/deploy-app.ts:205]: here you mentioned total pages 2 needed, but we passing `extraPages` as  1, where is another page.?

File: packages/runtime/test/integration/deploy-app.ts [hunk 5]
```diff
@@ -115,4 +115,148 @@ describe("Algorand Smart Contracts - Stateful Contract Account", function () {
...
+				appDefinition: {
+					appName: "App",
+					metaType: types.MetaType.FILE,
+					approvalProgramFilename: "very-long-approval-2-pages.teal",
+					clearProgramFilename: clearProgramFilename,
+					globalBytes: 1,
+					globalInts: 1,
+					localBytes: 1,
+					localInts: 1,
+					extraPages: 4 // should fail because extra pages range is [0, 3]
```
Comment (Megha-Dev-19) [inline packages/runtime/test/integration/deploy-app.ts:252]: I am having hard time understanding the pages here, could you help me by explaining it. Thanks

File: packages/runtime/src/ctx.ts (hunk 2 above)
Comment (rmeena840) [reply packages/runtime/src/ctx.ts:140]: No, it depends on the number of extra pages you have used. It's dynamic.
e.g extra page: 1, then limit is 4096
extra page: 2, then limit 6144

File: packages/runtime/test/integration/deploy-app.ts (hunk 3 above)
Comment (rmeena840) [reply packages/runtime/test/integration/deploy-app.ts:152]: Done

File: packages/runtime/test/integration/deploy-app.ts (hunk 4 above)
Comment (rmeena840) [reply packages/runtime/test/integration/deploy-app.ts:205]: Default total page is 1 and when you add extra page let's say 1 then total page count is 2.
So, it's. (default page(1) + extra page).

File: packages/runtime/test/integration/deploy-app.ts (hunk 5 above)
Comment (rmeena840) [reply packages/runtime/test/integration/deploy-app.ts:252]: check above comment

File: CHANGELOG.md [hunk 6]
```diff
@@ -51,6 +51,7 @@ from `algosdk` and sends it to the network.
 - Added additional checks for the `maxStackByteElementSize` and `maxProgramLength` in `Runtime`.
 - Added partial support for `bn254` opcodes. The full support delyed until the official release of `TEALv8/v9`.
 - Added a new pre-commit hook to check if there are any `.only`in tests . If there are any it will display error in which files/lines the problem was found.
+- Added program length check for deploy app on the basis the basis of extra pages in `rumtime`.
```
Comment (sczembor) [inline CHANGELOG.md:54]: something wrong here

File: packages/runtime/src/ctx.ts (hunk 2 above)
Comment (sczembor) [reply packages/runtime/src/ctx.ts:140]: so lets add the formula

File: CHANGELOG.md (hunk 6 above)
Comment (rmeena840) [reply CHANGELOG.md]: corrected it

**Comment by rmeena840** [comment]
> I leave some comments You also need to add changelog and also can you provide some description about this PR please You could link me to the docs about checking the length of the program too

@thdailong Done

File: packages/runtime/test/integration/deploy-app.ts (hunk 4 above)
Comment (sczembor) [reply packages/runtime/test/integration/deploy-app.ts:205]: then lets add this in the comment

File: packages/runtime/test/integration/deploy-app.ts [hunk 7]
```diff
@@ -115,4 +115,148 @@ describe("Algorand Smart Contracts - Stateful Contract Account", function () {
...
+					extraPages: 1 // should pass because total 2 pages needed
+				},
+				payFlags: {
+					totalFee: 1000,
+				},
+			};
+			assert.doesNotThrow(() => runtime.executeTx([execParams]))
+		});
+
+		it("Should fail: when sufficient extra pages was not defined", function () {
```
Comment (sczembor) [inline packages/runtime/test/integration/deploy-app.ts:214]:
```suggestion
		it("Should fail when sufficient extra pages was not defined", function () {
```

File: packages/runtime/src/ctx.ts (hunk 1 above)
Comment (rmeena840) [reply packages/runtime/src/ctx.ts:408]: Done

File: packages/runtime/src/ctx.ts (hunk 2 above)
Comment (rmeena840) [reply packages/runtime/src/ctx.ts:140]: done

**Review by sczembor — COMMENTED** [review]
good job, lets change one test case and pass it as clear program and lets remove the `:` from test description. Also I think it would be nice if the test would cover edge cases instead. So the one that should pass is exactly maxLen, and the other is maxLen +1 if this works we dont need to check anything else

File: packages/runtime/test/integration/deploy-app.ts (hunk 4 above)
Comment (rmeena840) [reply packages/runtime/test/integration/deploy-app.ts:205]: done

**Comment by rmeena840** [comment]
> good job, lets change one test case and pass it as clear program and lets remove the `:` from test description. Also I think it would be nice if the test would cover edge cases instead. So the one that should pass is exactly maxLen, and the other is maxLen +1 if this works we dont need to check anything else

@sczembor Done

File: packages/runtime/test/src/parser/parser.ts [hunk 8]
```diff
@@ -2724,5 +2724,23 @@ describe("Parser", function () {
...
+
+		it("Should pass when (program size + args size) = LogicSigMaxSize", function () {
+			const file = "test-arg.teal"; // byte size 3
+			interpreter.runtime = new Runtime([]);
+			interpreter.runtime.ctx.args = [new Uint8Array(997)];
+
+			assert.doesNotThrow(() => parser(getProgram(file), ExecutionMode.SIGNATURE, interpreter));
+		});
+
+		it("Should fail when (program size + args size) != LogicSigMaxSize", function () {
```
Comment (thdailong) [inline packages/runtime/test/src/parser/parser.ts:2736]:
```suggestion
		it("Should fail when (program size + args size) > LogicSigMaxSize", function () {
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -51,6 +51,8 @@ from `algosdk` and sends it to the network.
 - Added additional checks for the `maxStackByteElementSize` and `maxProgramLength` in `Runtime`.
 - Added partial support for `bn254` opcodes. The full support delyed until the official release of `TEALv8/v9`.
 - Added a new pre-commit hook to check if there are any `.only`in tests . If there are any it will display error in which files/lines the problem was found.
+- Added program length check on app deploy on the basis of extra pages in `rumtime`.
```
Comment (thdailong) [inline CHANGELOG.md:54]:
```suggestion
- Added program length check on app deploy on the basis of extra pages in `runtime`.
```

**Review by thdailong — APPROVED** [review]
Just left some suggestion

File: packages/runtime/src/parser/parser.ts [hunk 10]
```diff
@@ -860,28 +860,19 @@ export function assertMaxCost(
...
-// verify max length of TEAL code is within consensus parameters
-function _assertMaxLen(len: number, mode: ExecutionMode): void {
-	if (mode === ExecutionMode.SIGNATURE) {
+/**
+ * verify max size of lsig
+ * @param lsigProgramArgsSize lsig program and arugments size
+ * @param mode Execution mode - Signature (stateless) OR Application (stateful)
+ */
+function assertLogicMaxLen(lsigProgramArgsSize: number, mode: ExecutionMode): void {
+	if (mode === ExecutionMode.SIGNATURE && lsigProgramArgsSize > LogicSigMaxSize) {
```
Comment (sczembor) [inline packages/runtime/src/parser/parser.ts:869]: what does Args refers to?

Comment (sczembor) [reply packages/runtime/src/parser/parser.ts:869]: Do we check it both for args and length?

File: packages/runtime/src/parser/parser.ts [hunk 11]
```diff
@@ -894,7 +885,14 @@ function _assertMaxLen(len: number, mode: ExecutionMode): void {
 export function parser(program: string, mode: ExecutionMode, interpreter: Interpreter): Op[] {
 	const opCodeList: Op[] = [];
 	let counter = 0;
-
+	let lsigProgramArgsSize = Buffer.from(program, "base64").length;
```
Comment (sczembor) [inline packages/runtime/src/parser/parser.ts:888]: the size doesn't refer to compiled version?

File: packages/runtime/src/parser/parser.ts [hunk 12]
```diff
@@ -915,8 +913,6 @@ export function parser(program: string, mode: ExecutionMode, interpreter: Interp
 		assertMaxCost(interpreter.gas, mode);
 	}
 
-	// TODO: check if we can calculate length in: https://www.pivotaltracker.com/story/show/176623588
```
Comment (sczembor) [inline packages/runtime/src/parser/parser.ts:918]: I dont think it solves the problem since we do not compile it here

File: packages/runtime/src/runtime.ts [hunk 13]
```diff
@@ -1056,6 +1062,19 @@ export class Runtime {
 		}
 	}
 
+	/**
+	 * Verifies extra pages doesn't overflows
```
Comment (sczembor) [inline packages/runtime/src/runtime.ts:1066]:
```suggestion
	 * Verifies extra pages do not exceed the limit
```

File: packages/runtime/test/integration/deploy-app.ts [hunk 14]
```diff
@@ -115,4 +116,164 @@ describe("Algorand Smart Contracts - Stateful Contract Account", function () {
 			RUNTIME_ERRORS.GENERAL.APP_NAME_ALREADLY_USED
 		);
 	});
+
+	describe('Extra Pages', function () {
+		it("Should pass when program length doesn't exceeds total allowed program length", function () {
```
Comment (sczembor) [inline packages/runtime/test/integration/deploy-app.ts:121]:
```suggestion
		it("Should pass when program length doesn't exceed total allowed program length", function () {
```

File: packages/runtime/test/integration/deploy-app.ts [hunk 15]
```diff
@@ -115,4 +116,164 @@ describe("Algorand Smart Contracts - Stateful Contract Account", function () {
...
+				},
+				payFlags: {
+					totalFee: 1000,
+				},
+			};
+
+			expectRuntimeError(() => runtime.executeTx([execParams]), RUNTIME_ERRORS.TEAL.MAX_LEN_EXCEEDED);
+		});
+
+		it("Should pass when sufficient extra pages was defined", function () {
```
Comment (sczembor) [inline packages/runtime/test/integration/deploy-app.ts:192]:
```suggestion
		it("Should pass when sufficient extra pages were defined", function () {
```

File: packages/runtime/test/integration/deploy-app.ts [hunk 16]
```diff
@@ -115,4 +116,164 @@ describe("Algorand Smart Contracts - Stateful Contract Account", function () {
...
+				appDefinition: {
+					appName: "App",
+					metaType: types.MetaType.FILE,
+					approvalProgramFilename: "very-long-approval-2-pages.teal",
+					clearProgramFilename: clearProgramFilename,
+					globalBytes: 1,
+					globalInts: 1,
+					localBytes: 1,
+					localInts: 1,
+					extraPages: 1 // should pass because total 2 pages needed. total page = default page(1) + extra page
```
Comment (sczembor) [inline packages/runtime/test/integration/deploy-app.ts:206]: I think is misleading because there is not really default extra page. We are just allowed to deploy 2k and if we need more we define the "extra" page thus we can have 4k

File: packages/runtime/test/src/parser/parser.ts [hunk 17]
```diff
@@ -2724,5 +2725,23 @@ describe("Parser", function () {
 				RUNTIME_ERRORS.TEAL.MAX_COST_EXCEEDED
 			);
 		});
+
+		it("Should pass when (program size + args size) = LogicSigMaxSize", function () {
+			const file = "test-arg.teal"; // byte size 3
+			interpreter.runtime = new Runtime([]);
+			interpreter.runtime.ctx.args = [new Uint8Array(LogicSigMaxSize - 3)];
+
```
Comment (sczembor) [inline packages/runtime/test/src/parser/parser.ts:2733]: lets check for the length again(lsig)

**Review by sczembor — COMMENTED** [review]
I am not sure if the lsig length is correct here. I think we need a compiled version. Also check the comment about the `extraPages` I think the comments are not accurate

File: packages/runtime/src/parser/parser.ts (hunk 10 above)
Comment (rmeena840) [reply packages/runtime/src/parser/parser.ts:869]: args is arugments.
yes, we have to check for both args size and lsgi size.

File: packages/runtime/test/integration/deploy-app.ts (hunk 16 above)
Comment (rmeena840) [reply packages/runtime/test/integration/deploy-app.ts:206]: No, I don't think so. The default is 0 when extra pages is not defined.

File: packages/runtime/src/parser/parser.ts (hunk 10 above)
Comment (rmeena840) [reply packages/runtime/src/parser/parser.ts:869]: Please see here: https://github.com/algorand/go-algorand-sdk/blob/13da8508530b7dc2e9ae62766cccd0baaff0f4bd/logic/logic.go#L83

they have used the same logic.

File: packages/runtime/src/parser/parser.ts (hunk 11 above)
Comment (rmeena840) [reply packages/runtime/src/parser/parser.ts:888]: No, we don't need that.
I have checked how they have verified this case over here: Please see here: https://github.com/algorand/go-algorand-sdk/blob/13da8508530b7dc2e9ae62766cccd0baaff0f4bd/logic/logic.go#L83

They are using bytecode of program.

File: packages/runtime/src/parser/parser.ts (hunk 12 above)
Comment (rmeena840) [reply packages/runtime/src/parser/parser.ts:918]: Please see here: https://github.com/algorand/go-algorand-sdk/blob/13da8508530b7dc2e9ae62766cccd0baaff0f4bd/logic/logic.go#L83

We don't need compiled version.

File: packages/runtime/test/src/parser/parser.ts (hunk 17 above)
Comment (rmeena840) [reply packages/runtime/test/src/parser/parser.ts:2733]: done

File: packages/runtime/src/parser/parser.ts (hunk 11 above)
Comment (sczembor) [reply packages/runtime/src/parser/parser.ts:888]: The argument passed there is type `byte` and we use string. Lets ask on Algorand discord!

**Review by sczembor — APPROVED** [review]
Approval. Before merging please make sure about the compiled/noncompiled length on Algorand discord
